Modules/itertoolsmodule.c

```diff
@@ -3403,9 +3404,35 @@ count_nextlong(countobject *lz)
 static PyObject *
 count_next(countobject *lz)
 {
-    if (lz->cnt == PY_SSIZE_T_MAX)
-        return count_nextlong(lz);
-    return PyLong_FromSsize_t(lz->cnt++);
+    PyObject *returned;
+    Py_ssize_t cnt;
+
+    cnt = FT_ATOMIC_LOAD_SSIZE_RELAXED(lz->cnt);
+    for (;;) {
+        if (cnt == PY_SSIZE_T_MAX) {
+            /* slow mode */
+            Py_BEGIN_CRITICAL_SECTION(lz);
+            returned = count_nextlong(lz);
+            Py_END_CRITICAL_SECTION();
+            return returned;
+        }
+#ifdef Py_GIL_DISABLED
+        /* thread-safe fast version (increment by one).
+         * If lz->cnt changed between the pervious read and now,
+         * that means another thread got in our way. In this case,
+         * update cnt to new value of lz->cnt, and try again.
+         * Otherwise, (no other thread updated lz->cnt),
+         * atomically update lz->cnt with the incremented value and
+         * then return cnt (the previous value)
+         */
```

Comment:
The text is a bit confusing as written. We can work on making it more clear, but I'm not sure we need a comment here -- it's a pretty vanilla atomic compare-exchange loop.